pyomo/repn/plugins/nl_writer.py

```diff
@@ -1671,17 +1696,29 @@ def handle_product_node(visitor, node, arg1, arg2):
         mult = arg1[1]
         if not mult:
             # simplify multiplication by 0 (if arg2 is zero, the
-            # simplification happens implicitly when we evaluate the
-            # constant below)
+            # simplification happens when we evaluate the constant
+            # below).  Note that this is not IEEE-754 compliant, and
+            # will map 0*inf and 0*nan to 0 (and not to nan).  We are
+            # including this for backwards compatibility with the NLv1
+            # writer, but arguably we should deprecate/remove this
+            # "feature" in the future.
```

Comment:
The problem is *exactly* `x*log(x)`...  IDAES has models where this shows up during initialization (calculating physical properties of a multicomponent stream uses `f[i] * fcn(T)*log(x[i])`, where `x[i] == f[i] / sum(f)` (that is, the molar flow multiplied by the molar concentration).  If the model is initialized where `x[i] == 0` (i.e., the component is not present - an arguably physically reasonable thing to do), it would be nice if the model were still solvable.  The limit of `x*log(x)` as x -> 0 is 0, so returning 0 is actually not insane (from the point of view of someone modeling a physical system).

Reply:
As an IDAES Developer, I am somewhat torn on this one - years of experience tell me that setting a concentration to exactly 0 is never a good idea (due to issues like this).

On the grounds of encouraging modelers to actually think about what they are doing, one part of me says that this should raise an exception (although from the IDAES perspective I would want the error to be context specific, i.e. don't set concentrations to zero, which is not appropriate in the general sense for Pyomo).

On the other hand, someone will always try to do this, and sometimes the path of least resistance is to let them get away with it (but it will probably come back and bite them later, in which case it is just delaying the inevitable).

Reply:
I think we should definitely toss a warning here.  I am torn if it should be a "buyer beware" warning, or a full deprecation warning (i.e., buyer beware, and we will rip the rug out from under you soon).

Reply:
@andrewlee94, how hard would it be to update IDAES property packages to not use `f[i] * fcn(T)*log(x[i])`?

@emma58 What do you mean by "the wrong answer"?  If we remove this special casing, models with `0*nan` or `0*inf` will have nan's in the LP/NL/whatnot file, resulting in an unintelligible solver parse error.  Or are you suggesting that we raise an exception for all nans?  What about `Expr_if(IF_=x>=0, y, nan)`?

Reply:
@jsiirola My guess is fairly hard - for one, what would you suggest it by replaced by? For better or worse, that is the generally accepted form for many properties, and I am not sure how amenable to reformulation it would be.

In many ways (and specifically because of this), I generally suggest not setting mole fractions to zero (and giving them a strictly positive domain).

Reply:
@jsiirola, I think you can get wrong answers if you *do* set `0*nan = 0` and `0*inf = 0` because models with undefined terms could suddenly become linear and friendly. So the right "answer" I think would be the solver screaming about nan's. So yes, I guess in that world order, we would just need to complain about nan's right away rather than writing them. For the `Expr_if` case, I feel like it should still be undefined (we should still complain) because it's unclear if that implicitly means that `x < 0` should be infeasible or what. It's ambiguous what the feasible region is, so how would you optimize over it?

Reply:
OK - this PR now includes a deprecation warning whenever `0*nan` is encountered.  We will probably want to remove the warning and the feature relatively quickly.